gateway/src/apicast/policy/rate_limiting_to_service/rate_limiting_to_service.lua

```diff
@@ -0,0 +1,201 @@
+  if not status then
+    catch_f(exception)
+  end
+end
+
+function _M:access()
+  local limitters = {}
+  local keys = {}
+  local states = {}
+
+  local limitters_limit_conn = {}
+  local keys_limit_conn = {}
+
+  local limitters_limit_conn_committed = {}
+  local keys_limit_conn_committed = {}
+
+  if not self.redis_info then
+    ngx.log(ngx.ERR, "No Redis information.")
+    return ngx.exit(500)
+  end
+
+  for _, limitter in ipairs(self.limitters) do
+    local limit
+    local class_not_found = false
+    try(
+      function()
+        limit = require (limitter.limitter)
+      end,
+      function(e)
+        ngx.log(ngx.ERR, "failed to find module: ", e)
+        class_not_found = true
+      end
+    )
+    if class_not_found then
+      return ngx.exit(500)
+    end
+
+    local lim, limerr
+    local failed_to_instantiate = false
+    try(
+      function()
+        lim, limerr = limit.new(shdict_key, unpack(limitter.values))
+        if not lim then
+          ngx.log(ngx.ERR, "failed to instantiate limitter: ", limerr)
+          failed_to_instantiate = true
+        end
+      end,
+      function(e)
+        ngx.log(ngx.ERR, "failed to instantiate limitter: ", e)
+        failed_to_instantiate = true
+      end
+    )
+    if failed_to_instantiate then
+      return ngx.exit(500)
+    end
+
+    local rediserr
+    lim.dict, rediserr = redis_shdict(self.redis_info.host, self.redis_info.port, self.redis_info.db)
+    if not lim.dict then
+      ngx.log(ngx.ERR, "failed to connect Redis: ", rediserr)
+      return ngx.exit(500)
+    end
+
+    limitters[#limitters + 1] = lim
+    keys[#keys + 1] = limitter.key
+
+    if limitter.limitter == "resty.limit.conn" then
+      limitters_limit_conn[#limitters_limit_conn + 1] = lim
+      keys_limit_conn[#keys_limit_conn + 1] = limitter.key
+    end
+  end
+
+  local delay, comerr = limit_traffic.combine(limitters, keys, states)
+  if not delay then
+    if comerr == "rejected" then
+      ngx.log(ngx.ERR, "Requests over the limit.")
+      return ngx.exit(429)
+    end
+    ngx.log(ngx.ERR, "failed to limit traffic: ", comerr)
+    return ngx.exit(500)
+  end
+
+  for i, lim in ipairs(limitters_limit_conn) do
+    if lim:is_committed() then
+      limitters_limit_conn_committed[#limitters_limit_conn_committed + 1] = lim
+      keys_limit_conn_committed[#keys_limit_conn_committed + 1] = keys_limit_conn[i]
+    end
+  end
+
+  if next(limitters_limit_conn_committed) ~= nil then
+    local ctx = ngx.ctx
+    ctx.limitters = limitters_limit_conn_committed
+    ctx.keys = keys_limit_conn_committed
+  end
+
+  if delay >= 0.001 then
+    ngx.log(ngx.WARN, 'need to delay by: ', delay, 's')
+    ngx.sleep(delay)
+  end
+
+end
+
+local function checkin(_, ctx, time, semaphore, redis_info)
+  local limitters = ctx.limitters
+  local keys = ctx.keys
+
+  for i, lim in ipairs(limitters) do
+    local rediserr
+    lim.dict, rediserr = redis_shdict(redis_info.host, redis_info.port, redis_info.db)
```

Comment:
This shouldn't be needed as the limiters are already initialized and have correct redis object in the `.dict` property, right?